R/ns.R

```diff
@@ -95,21 +95,41 @@ step_ns_new <-
     )
   }
 
-
 ns_wrapper <- function(x, args) {
-  if (!("Boundary.knots" %in% names(args)))
-    args$Boundary.knots <- range(x)
-  args$x <- x
-  ns_obj <- do.call("ns", args)
-  ## don't need to save the original data so keep 1 row
-  out <- matrix(NA, ncol = ncol(ns_obj), nrow = 1)
+  # Only do the parameter computations from splines::bs() / splines::ns(), don't evaluate at x.
+  degree <- 1L
+  intercept <- as.logical(args$intercept %||% FALSE)
+  # This behaves differently from splines::ns() if length(x) is 1
+  boundary <- sort(args$Boundary.knots) %||% range(x)
+
+  # This behaves differently from splines::bs() and splines::ns() if num_knots < 0L
+  # the original implementations issue a warning.
+  if (!is.null(args$df) && is.null(args$knots) && args$df - degree - intercept >= 1L) {
+    num_knots <- args$df - degree - intercept
+    ok <- !is.na(x) & x >= boundary[1L] & x <= boundary[2L]
+    knots <- unname(quantile(x[ok], seq_len(num_knots) / (num_knots + 1L)))
+  } else {
+    knots <- numeric()
+  }
+
+  # Only construct the data necessary for splines_predict
+  out <- matrix(NA, ncol = degree + length(knots) + intercept, nrow = 1L)
   class(out) <- c("ns", "basis", "matrix")
-  attr(out, "knots") <- attr(ns_obj, "knots")[]
-  attr(out, "Boundary.knots") <- attr(ns_obj, "Boundary.knots")
-  attr(out, "intercept") <- attr(ns_obj, "intercept")
+  attr(out, "knots") <- knots
```

Comment:
Why are the attributes copied back to the object? These columns are added to the data frame and they are stripped off there.

Reply:
This can't be removed as they are needed for the predict step.
Commenting out these lines for `bs` errors the tests like this:

```
-- Error (<text>:19:3): correct basis functions --------------------------------
Error: unused arguments (`NA` = NULL, `NA` = NULL, `NA` = NULL, `NA` = NULL)
Backtrace:
 1. recipes::prep(with_bs, training = biomass_tr, verbose = FALSE)
 2. recipes::prep.recipe(with_bs, training = biomass_tr, verbose = FALSE) R/recipe.R:306:8
 4. recipes::bake.step_bs(x$steps[[i]], new_data = training) R/recipe.R:520:8
 5. recipes::bs_predict(...) R/bs.R:174:4
 7. splines:::predict.bs(object, xu)
 8. base::do.call("bs", a)

-- Error (<text>:59:3): printing -----------------------------------------------
Error: unused arguments (`NA` = NULL, `NA` = NULL, `NA` = NULL, `NA` = NULL)
Backtrace:
  1. testthat::expect_output(...)
 11. recipes::prep.recipe(with_bs, training = biomass_tr, verbose = TRUE) R/recipe.R:306:8
 13. recipes::bake.step_bs(x$steps[[i]], new_data = training) R/recipe.R:520:8
 14. recipes::bs_predict(...) R/bs.R:174:4
 16. splines:::predict.bs(object, xu)
 17. base::do.call("bs", a)

Test passed 
```

and `ns` tests fail similarly:

```
-- Error (<text>:19:3): correct basis functions --------------------------------
Error: unused arguments (`NA` = NULL, `NA` = NULL, `NA` = NULL)
Backtrace:
 1. recipes::prep(with_ns, training = biomass_tr, verbose = FALSE)
 2. recipes::prep.recipe(with_ns, training = biomass_tr, verbose = FALSE) R/recipe.R:306:8
 4. recipes::bake.step_ns(x$steps[[i]], new_data = training) R/recipe.R:520:8
 5. recipes::ns_predict(...) R/ns.R:167:4
 7. splines:::predict.ns(object, xu)
 8. base::do.call("ns", a)

-- Error (<text>:59:3): printing -----------------------------------------------
Error: unused arguments (`NA` = NULL, `NA` = NULL, `NA` = NULL)
Backtrace:
  1. testthat::expect_output(...)
 11. recipes::prep.recipe(with_ns, training = biomass_tr, verbose = TRUE) R/recipe.R:306:8
 13. recipes::bake.step_ns(x$steps[[i]], new_data = training) R/recipe.R:520:8
 14. recipes::ns_predict(...) R/ns.R:167:4
 16. splines:::predict.ns(object, xu)
 17. base::do.call("ns", a)

Test passed 
```